crates/sui-types/src/dynamic_field.rs

```diff
@@ -1,27 +1,49 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
+use crate::base_types::ObjectDigest;
+use crate::error::{SuiError, SuiResult};
+use crate::sui_serde::Readable;
+use crate::{ObjectID, SequenceNumber, SUI_FRAMEWORK_ADDRESS};
 use move_core_types::language_storage::{StructTag, TypeTag};
 use move_core_types::value::{MoveStruct, MoveValue};
 use schemars::JsonSchema;
 use serde::Deserialize;
 use serde::Serialize;
-
-use crate::base_types::ObjectDigest;
-use crate::error::{SuiError, SuiResult};
-use crate::{ObjectID, SequenceNumber, SUI_FRAMEWORK_ADDRESS};
-
-#[derive(Clone, Serialize, Deserialize, JsonSchema, Ord, PartialOrd, Eq, PartialEq, Debug)]
+use serde_json::Value;
+use serde_with::serde_as;
+use serde_with::DisplayFromStr;
+use std::fmt::{Display, Formatter};
+#[derive(Clone, Serialize, Deserialize, JsonSchema, Debug)]
 #[serde(rename_all = "camelCase")]
 pub struct DynamicFieldInfo {
-    pub name: String,
+    pub name: DynamicFieldName,
     pub type_: DynamicFieldType,
     pub object_type: String,
     pub object_id: ObjectID,
     pub version: SequenceNumber,
     pub digest: ObjectDigest,
 }
 
+#[serde_as]
+#[derive(Clone, Serialize, Deserialize, JsonSchema, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct DynamicFieldName {
+    #[schemars(with = "String")]
+    #[serde_as(as = "Readable<DisplayFromStr, _>")]
+    pub type_: TypeTag,
+    // Bincode does not like serde_json::Value, rocksdb will not insert the value without serializing value as string.
+    #[schemars(with = "Value")]
+    #[serde_as(as = "Readable<_, DisplayFromStr>")]
+    pub value: Value,
```

Comment:
It seems a bit json-rpc specific to use a `json::Value` here--someone looking at a `DynamicFieldName` from inside Rust would presumably want to see a `MoveStruct` or something else a bit more structured

Reply:
I thought MoveStruct is quite hard to construct in rust. Currently user can use the serde_json's macro `json!{}` to construct the value, it is not structured but quite convenient. Maybe I can add something in the Rust SDK to accept more structured MoveValue input.